src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Diagnostics/BindingFailedEventArgs.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+//
+// Description:
+//      Contains EventArg types raised to communicate BindingFailed events.
+
+using System.Diagnostics;
+using System.Windows.Data;
+
+namespace System.Windows.Diagnostics
+{
+    /// <summary>
+    /// Provides data for <see cref="BindingDiagnostics.BindingFailed"/> 
+    /// </summary>
+    public class BindingFailedEventArgs : EventArgs
+    {
+        /// <summary>
+        /// For filtering failures (warnings, errors, etc).
+        /// </summary>
+        public TraceEventType EventType { get; }
+
+        /// <summary>
+        /// Failure code.
+        /// </summary>
+        public int Code { get; }
+
+        /// <summary>
+        /// This is the full message that is also written to debug output.
+        /// </summary>
+        public string Message { get; }
+
+        /// <summary>
+        /// Can be null for some failure codes that don't have a BindingExpressionBase context.
+        /// </summary>
+        public BindingExpressionBase Binding { get; }
```

Comment:
Lifetime problem:  A listener that retains a reference to the Binding beyond the scope of the event will affect the lifetime of the binding object and everything it refers to, which could encompass a lot - visual subtrees, etc..  This could change the behavior of the app, in obvious ways like perceived memory leaks, and in less obvious ways like bindings and their target elements remaining active after the app thinks they're gone.  [This applies to your "pending event" list as well.]

How is VS planning to use Binding?  In particular, will you need it after the event, say to help a user who asks about a binding error that happened long ago and whose binding wouldn't exist any more sans BindingDiagnostics?  If so, we can't fix the problem, but we (well, you - VS) should warn users that "running with BindingDiagnostics enabled can change the behavior of the app, take appropriate precautions".   If not, i.e. if your event listeners will harvest what they need from the Binding, is there some way to put the harvested data in the event args?

In any case, please add a comment here to warn other providers of diagnostic assistants of the issue.

Reply:
Actually all VS does is harvest some strings and source info from the Binding, nothing is cached. The pending event list will hold onto Bindings until VS has a chance to flush the events. So that will prevent garbage collection. Seems like my options are:

1. Not cache the Binding, just cache everything I need from it. Hopefully VS doesn't need more info in the future
2. Use a WeakReference instead, and hope that Bindings aren't garbage collected before the pending events are flushed
3. Keep as is, but add more comments and documentation about how objects may have a long lifetime with diagnostics enabled

I'll see exactly what change I'd need for #1 (remove the Binding reference and cache data in here rather than VS code)

---
In reply to: [488330836](https://github.com/dotnet/wpf/pull/3505#discussion_r488330836) [](ancestors = 488330836)

Reply:
@SamBent I have the change locally to do all the binding harvesting in the event args. It seems good to me, and I've been working with data binding failures enough to be confident that the harvested data won't need to change (at least for Visual Studio's feature). I'm tempted to put a WeakReference to the binding just in case...

Reply:
@spadapet I'm actually fine with (3), given that you aren't actually caching the bindings.  It might be more flexible if we decide to expose this to more tools.  

A WeakReference doesn't help much, as nothing prevents the listener from caching its target.  And I'm not too worried about the pending events, given your claim that the listener will arrive as soon as it can to collect them.   All the diagnostic features rely on the good will of the listeners anyway, so adding to the "good will" list is reasonable.  Mostly I wanted to know that you weren't (inadvertently) breaking that good will, and get you to document what "good will" means for this event, so that other tools will know how to behave.

Reply:
@SamBent Thanks, I'll stick with (1) then, and add a comment here. The diagnostics features already allow the app to be broken by messing with property values, so maybe there is already a warning.